### PR TITLE
🚧 Report workspace button refresh

### LINKED ISSUE #2982: Switching workspaces in Reports is not updated after sync
The floating workspace switching button does not change when user gains or loses access to a workspace and it shows the option to switch to the previously known workspaces.

It is not shown or hidden when the number of workspaces increases from 1 to 2 or more or decreeses back to 1.

If a workspace which is no longer accessible was selected before, it will be still shown and used to fetch reports (although the reports won't be loaded anymore).

## Context
The floating button was added in #2803 

## Expected behavior
When the list of available workspaces changes, the button is ReportsViewModel adjusts its state accordingly and visual elements are shown or hidden correctly with up-to-date data.

## Suggestions
We need to change the workspace data source to an observable data source and observe the changes of workspaces. Don't forget that workspaces which are marked as ghosts should not be accessible as well.

### PR DESCRIPTION
**Pull request by sofokli1**
## What's this?
This makes the switch workspace button show the changes in the workspaces after sync and hides the button if there is only 1 workspace.

### Relationships
Closes #2982

## Why do we want this?
We want the button and what it shows to update as the data changes.

## How is it done?
Partial rxfactoring of the iOS VM/VC and change of the workspace from datasource to observableDatasource.

### Why not another way?
Can't think of any

### Side effects
Only functional tests can tell

## Review hints
Just test a lot. When leaving a workspace in web it never gets removed in the app, not sure why but i don't think this is introduced in this PR. Also i am having trouble with 1 test, even thought the data are correct it fails 😤 crazy stuff (help is very welcome)

## :squid: Permissions
:squid:

### CODE REVIEW

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/ReportsViewModelTests.cs [hunk 1]
```diff
@@ -509,14 +527,22 @@ public async Task ShouldTriggerAReportReload()
...
                 await ViewModel.Initialize();
                 var mockWorkspace = new MockWorkspace { Id = WorkspaceId + 1, Name = "Selected workspace" };
                 DialogService.Select(Arg.Any<string>(), Arg.Any<IEnumerable<(string, IThreadSafeWorkspace)>>())
                     .Returns(Observable.Return(mockWorkspace));
 
-                await ViewModel.SelectWorkspace.ExecuteAsync();
+                await ViewModel.SelectWorkspaceAction();
 
-                ViewModel.WorkspaceName.Should().Be(mockWorkspace.Name);
+                TestScheduler.Start();
```
Comment (khoi) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/ReportsViewModelTests.cs]: This is the failing test I assume? 

The problem is the `observer.Messages` value is `"", null, "Selected Workspace"`.

Notice how the second value is null, this value is emitted in `ReportViewModel`'s `Initialize

```csharp
public override async Task Initialize()
{ 
     var workspace = await interactorFactory.GetDefaultWorkspace().Execute();
     workspaceNameSubject.OnNext(workspace.Name);
}
```

In the test, this is not mocked which lead to the null value.

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/ReportsViewModelTests.cs [hunk 2]
```diff
@@ -509,14 +527,22 @@ public async Task ShouldTriggerAReportReload()
...
+                observer.Messages.AssertEqual(
+                    ReactiveTest.OnNext(1, ""),
```
Comment (khoi) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/ReportsViewModelTests.cs]: This is handy, I didn't know that you could do this 👍

File: Toggl.Daneel/ViewControllers/ReportsViewController.cs [hunk 3]
```diff
@@ -50,42 +51,32 @@ public override void ViewDidLoad()
             source.OnScroll += onReportsTableScrolled;
             ReportsTableView.Source = source;
 
-            var areThereEnoughWorkspaces = new LambdaConverter<ICollection<(string, IThreadSafeWorkspace)>, bool>(workspaces => workspaces.Count > 1);
```
Comment (simonrozsival) [inline Toggl.Daneel/ViewControllers/ReportsViewController.cs]: We can now also 🔥 the `LambdaConverter` itself. Please delete it so we don't forget about it and don't have the unnecessary class in the codebase.

File: Toggl.Foundation.MvvmCross/ViewModels/ReportsViewModel.cs [hunk 4]
```diff
@@ -54,6 +56,9 @@ public sealed class ReportsViewModel : MvxViewModel
         private DateFormat dateFormat;
         private IReadOnlyList<ChartSegment> segments = new ChartSegment[0];
         private IReadOnlyList<ChartSegment> groupedSegments = new ChartSegment[0];
+        private readonly BehaviorSubject<ICollection<(string ItemName, IThreadSafeWorkspace Item)>> workspacesSubject = new BehaviorSubject<ICollection<(string ItemName, IThreadSafeWorkspace Item)>>(new Collection<(string ItemName, IThreadSafeWorkspace Item)>());
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/ReportsViewModel.cs]: 🔥

File: Toggl.Foundation.MvvmCross/ViewModels/ReportsViewModel.cs [hunk 5]
```diff
@@ -38,6 +39,7 @@ public sealed class ReportsViewModel : MvxViewModel
         private readonly IInteractorFactory interactorFactory;
         private readonly IAnalyticsService analyticsService;
         private readonly IDialogService dialogService;
+        private readonly ISchedulerProvider schedulerProvider;
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/ViewModels/ReportsViewModel.cs]: The `schedulerProvider` is used only in the constructor, we don't need the reference to it outside of it and so we don't have to store it in a field. 🔥 please.

File: Toggl.Foundation.MvvmCross/ViewModels/ReportsViewModel.cs [hunk 6]
```diff
@@ -360,14 +401,14 @@ private IReadOnlyList<ChartSegment> groupSegments()
                 .AsReadOnly();
         }
 
-        private async Task selectWorkspace()
+        public async Task SelectWorkspaceAction()
```
Comment (arturs105) [inline Toggl.Foundation.MvvmCross/ViewModels/ReportsViewModel.cs]: Why add the "action" part to the name? 🤔

File: Toggl.Daneel/ViewControllers/ReportsViewController.cs [hunk 7]
```diff
@@ -50,42 +51,32 @@ public override void ViewDidLoad()
...
                 return size.Width >= maximumWorkspaceNameLabelWidth;
-            });
+            };
+
+            //Text
+            this.Bind(ViewModel.WorkspaceNameObservable, WorkspaceLabel.BindText());
+            this.Bind(ViewModel.CurrentDateRangeStringObservable, titleButton.BindTitle());
+
+            //Visibility
+            this.Bind(ViewModel.WorkspacesObservable.Select(areThereEnoughWorkspaces), WorkspaceButton.BindIsVisible());
```
Comment (simonrozsival) [inline Toggl.Daneel/ViewControllers/ReportsViewController.cs]: When you open the reports view with a single workspace, the button is visible for a brief moment and then it is hidden. An easy fix: add `.StartWith(false)` to the observable.

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/ReportsViewModelTests.cs [hunk 8]
```diff
@@ -266,14 +269,22 @@ public async Task IsInitializedToEmptyOrNull()
...
+
                 var currentDate = new DateTimeOffset(currentYear, currentMonth, currentDay, 0, 0, 0, TimeSpan.Zero);
                 var start = new DateTimeOffset(startYear, startMonth, startDay, 0, 0, 0, TimeSpan.Zero);
                 var end = new DateTimeOffset(endYear, endMonth, endDay, 0, 0, 0, TimeSpan.Zero);
                 TimeService.CurrentDateTime.Returns(currentDate);
                 ViewModel.ChangeDateRangeCommand.Execute(
                     ReportsDateRangeParameter.WithDates(start, end).WithSource(ReportsSource.Calendar));
 
-                ViewModel.CurrentDateRangeString.Should().Be($"{Resources.ThisWeek} ▾");
+                //ViewModel.CurrentDateRangeString.Should().Be($"{Resources.ThisWeek} ▾");
```
Comment (arturs105) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/ReportsViewModelTests.cs]: This should be either uncommented or removed.

File: Toggl.Daneel/ViewControllers/ReportsViewController.cs (hunk 7 above)
Comment (sofokli1) [reply Toggl.Daneel/ViewControllers/ReportsViewController.cs]: What do you think about having the `WorkspaceButton` hidden in IB by default?

File: Toggl.Foundation.MvvmCross/ViewModels/ReportsViewModel.cs (hunk 6 above)
Comment (sofokli1) [reply Toggl.Foundation.MvvmCross/ViewModels/ReportsViewModel.cs]: there is something else called `SelectWorkspace`. Will rename `SelectWorkspaceAction` to `SelectWorkspaceMethod`. How does this sound?

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/ReportsViewModelTests.cs (hunk 1 above)
Comment (sofokli1) [reply Toggl.Foundation.Tests/MvvmCross/ViewModels/ReportsViewModelTests.cs]: thanx, I was going crazy with this one 😤

**Comment by sofokli1** [comment]
@khoi since the app does not detect when it loses access to a workspace there isn't much to test in that situation until that feature is available.

File: Toggl.Daneel/ViewControllers/ReportsViewController.cs (hunk 7 above)
Comment (simonrozsival) [reply Toggl.Daneel/ViewControllers/ReportsViewController.cs]: That would work as well, I am fine with that.

**Pull request merged**